docs/design/proposals/resource-reference/resource-reference.md

```diff
@@ -0,0 +1,555 @@
+
+* This new spec field will work as additional input for an existing spec field.
+Addition of the extra spec field keeps the backward compatibility with existing
+custom resource definitions.
+
+* Validation will be present that only one of existing spec field OR new reference
+field is present in custom resource, not both.
+
+* If the existing spec field was marked as required, validation will be added that
+at least one of existing spec field or new reference field is present.
+
+* If `AWSResourceReferenceWrapper` type field is present in the desired state,
+first ACK controller will query for the referenced resource. If the referenced resource
+has `ACK.ResourceSynced` condition status `True` & specified field is found
+in the referenced resource, the reconciliation loop will progress. Otherwise
+the request will get requeued for future processing.
+
+* Watchers for the referenced resource type will not be implemented in the initial
+implementation. The initial implementation will mostly be used for finding
+identifier of referenced resource and any update in referenced resource should not
+change it's identifier. As mentioned earlier, deletion will also be handled in
+future.
+
+* The code to determine when the referenced resource is synced will be generated,
+and additional hooks will be added for customization. Ex: Do not just check presence
+of a field but also check if the Status is available etc...
+
+### Low Level Implementation
+
+#### generator.yaml Changes
+
+```go
+type FieldConfig struct {
+    ...
+    References *ReferencesConfig `json:"references,omitempty"`
+    ...
+}
+
+// ReferencesConfig contains the instructions for how to add the referenced resource
+// configuration for a field.
+type ReferencesConfig struct {
+	// Resource mentions the K8s resource which is read to resolve the
+	// reference
+	Resource string `json:"resource"`
+	// Path refers to the the path of field which should be copied
+	// to resolve the reference
+	Path string `json:"path"`
+}
+```
+
+`APIGatewayv2 generator.yaml` Example:
+```yaml
+resources:
+  Integration:
+    fields:
+      ApiId:
+        references:
+          resource: API
+          path: Status.APIID
+```
+
+#### New AWSResourceReference Types
+
+```go
+// AWSResourceReferenceWrapper provides a wrapper around *AWSResourceReference
+// type to provide more user friendly syntax for references using 'from' field
+// Ex:
+// APIIDRef:
+//   from:
+//     name: my-api
+type AWSResourceReferenceWrapper struct {
+	From *AWSResourceReference `json:"from,omitempty"`
+}
+
+// AWSResourcesReferenceWrapper provides a wrapper around slice of *AWSResourceReference
+// type to provide more user friendly syntax for references using 'from' field
+// Ex:
+// SecurityGroupIDs:
+//   from:
+//     - name: my-sg-1
+//     - name: my-sg-2
+type AWSResourcesReferenceWrapper struct {
+	From []*AWSResourceReference `json:"from,omitempty"`
+}
```

Comment:
I would have expected slice of identifier fields to use `[]*AWSResourceReferenceWrapper`:

```yaml
SecurityGroupIDs:
  - from:
      name: my-sg-1
  - from:
      name: my-sg-2
```

Reply:
I was hoping to avoid the many 'from' keyword, But above works as well.
It can be done either way i.e. `[]*AWSResourceReference` or `[]*AWSResourceReferenceWrapper`

Reply:
I agree that the wrapper structure is ugly, but from the the YAML output I prefer the single `from` keyword that @vijtrip2 is assuming. `from` is there to signify that it's a reference, rather than a value. Multiple `from` doesn't add any extra value to that IMO.

Reply:
@RedbackThomson the problem with using the wrapper like that is the plural property of the source field is lost. In other words, the fact that the field stores *multiple* identifiers or a single identifier is lost from the schema of the resource itself.

Reply:
Ah totally valid, then. You're using the `from` to identify a single reference. Thanks for the explanation.